RFC-0034-batch-kronecker-product.md

```diff
@@ -0,0 +1,170 @@
+
+### In use
+For a given dimension along which to batch, the requirement is that all dimensions agree up to and including the batch dimension. For example,
+```
+a.shape = (2,3,4)
+b.shape = (2,3,5)
+```
+could be batched along dimension 0 and 1, whereas
+```
+a.shape = (3,7,2)
+b.shape = (3,5,3)
+```
+could only be batched along dimension 0.
+
+Some more examples:
+
+```
+a.shape = (2,5,4)
+b.shape = (2,5,7)
+batch_kron(a, b, 0).shape = (2,25,28)
+batch_kron(a, b, 1).shape = (2,5,28)
+
+a.shape = (5,4,3)
+b.shape = (5,3,9)
+batch_kron(a, b, 0).shape = (5, 12, 27)
+batch_kron(a, b, 1) 'Error: required dimensions do not match.'
+```
+For all dimensions higher than the batch dimension, the regular Kronecker product is applied to each pair of tensors. The result is a tensor of the required shape, as seen in the next subsection.
+
+### How it works internally
+The method is similar to how `kron` currently works within LibTorch, although extra care has to be taken when handling the batch dimensions. 
+
+The idea is to reshape the two tensors in such a way that there are dimensions of size one inserted at alternating points in the shape of each tensors. This means that when the vectors are Kroneckered, the dimensions align accordingly and the result can be reshaped as needed.
+
+For example, consider 
+```
+a.shape = (2,3,5)
+b.shape = (2,3,8)
+```
+For the regular Kronecker product, the reshaping looks like the following:
+```
+a_reshape = (2,1,3,1,5,1)
+b_reshape = (1,2,1,3,1,8)
+result_reshape = (4,9,40)
+```
+We see that the resultant shape is merely an element-wise product of the initial shapes.
+
+For the batch Kronecker product (in dimension zero), we have the following:
+
+```
+a_reshape = (2,3,1,5,1)
+b_reshape = (2,1,3,1,8)
+result_reshape = (2, 9, 40)
+```
+
+And in dimension one:
+
+```
+a_reshape = (2,3,5,1)
+b_reshape = (2,3,1,8)
+result_reshape = (2,3,40)
+```
+
+We can see that for the dimensions up to and including the batch dimension, there is no change; and that the dimensions after the batch dimension are alternated in the same manner as the regular Kronecker product. 
+
+The element-wise multiplying of the dimension sizes in the resultant shape also follows this parallel to the regular Kronecker product.
+
+
+### A further generalisation
+
+This functionality is beyond what I need, but I feel that this generalisation would be suited to PyTorch.
+
+So far, all of my uses have been with tensors that have an equal number of dimensions. But there is nothing to say this must always be the case, since the Kronecker product will work will tensors of any shape.
+
+For example,
+```
+a.shape = (2,5,2,3)
+b.shape = (2,5,6)
+batch_kron(a, b, 0).shape = (2,5,10,18)
+batch_kron(a, b, 1).shape = (2,5,2,18)
+```
+We see that after the batch dimensions, the rest of the dimensions are right-aligned and Kroneckered as usual.
+
+
+## **Drawbacks**
+This is not a breaking change.
```

Comment:
Adding this new kwarg to `torch.kron` wouldn't be bc-breaking either but would make it diverge from numpy's kron.
cc @rgommers any plans on adding kron to the arrray API that we should follow here?